### PR TITLE
MultiGPU training + changes to Checkpointing logic

### PR DESCRIPTION
**Pull request by prabhuteja12**
This is WIP to rework checkpointing and the multiGPU training in Renate. 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by prabhuteja12** [comment]
There are some clean-ups that are not part of this PR. 

* Most of the `set_*` methods can be eliminated. We no longer use these to restore checkpoints. 
* `components` need to be handled with a little more care. See #236 .

File: src/renate/utils/misc.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+from typing import Union
+from pathlib import Path
+
+from pytorch_lightning.utilities.rank_zero import rank_zero_only
+
+
+def int_or_str(x: str) -> Union[str, int]:
```
Comment (wistuba) [inline src/renate/utils/misc.py:9]: not sure we have a good location for this function but please check for type of x instead of try expect casting

File: test/conftest.py [hunk 2]
```diff
@@ -254,6 +270,11 @@ def get_renate_module_mlp(
     )
 
 
+@pytest.helpers.register
+def get_loss_fn() -> torch.nn.Module:
```
Comment (wistuba) [inline test/conftest.py:274]: can we use this function in the cases above? does that make sense?

File: test/renate/benchmark/test_experimentation.py [hunk 3]
```diff
@@ -19,8 +19,10 @@ def experiment_job_kwargs():
         "mode": "max",
         "metric": "val_accuracy",
         "num_updates": 2,
-        "max_time": 15,
+        "max_time": 360,
```
Comment (wistuba) [inline test/renate/benchmark/test_experimentation.py:22]: undo this change, this was done for debugging afaik

File: test/renate/updaters/avalanche/test_avalanche_model_updater.py [hunk 4]
```diff
@@ -26,7 +26,7 @@ def get_model_and_dataset(dataset_size: int = 100):  # TODO: remove
 
 @pytest.mark.parametrize("provide_folder", [True, False])
 def test_avalanche_model_updater(tmpdir, provide_folder):
-    model, train_dataset, test_data = pytest.helpers.get_renate_module_mlp_and_data(
+    model, train_dataset, test_data, _ = pytest.helpers.get_renate_module_mlp_data_and_loss(
```
Comment (wistuba) [inline test/renate/updaters/avalanche/test_avalanche_model_updater.py:29]: why do we use the "and loss" function and then drop the loss?

File: test/renate/updaters/experimental/test_er.py [hunk 5]
```diff
@@ -91,8 +92,19 @@ def test_er_validation_buffer(tmpdir):
...
     [
-        [ExperienceReplayLearner, {"alpha": 0.2, "memory_size": 10, "memory_batch_size": 10}],
-        [DarkExperienceReplayLearner, {"alpha": 0.1, "beta": 0.3, "memory_size": 42}],
+        [
+            ExperienceReplayLearner,
+            {
+                "alpha": 0.2,
+                "memory_size": 10,
+                "memory_batch_size": 10,
+                "loss_fn": torch.nn.CrossEntropyLoss(),
```
Comment (wistuba) [inline test/renate/updaters/experimental/test_er.py:101]: when do we use the pytest.helpers.get_loss_fn?

File: test/renate/updaters/experimental/test_fine_tuning.py [hunk 6]
```diff
@@ -20,13 +20,23 @@ def get_model_and_dataset():
     return model, dataset
 
 
-def test_fine_tuning_updater():
+@pytest.mark.parametrize("devices", [None, 1])
+@pytest.mark.parametrize("strategy", [None, "ddp"])
+@pytest.mark.parametrize("accelerator", ["cpu"])
+def test_fine_tuning_updater(devices, strategy, accelerator):
```
Comment (wistuba) [inline test/renate/updaters/experimental/test_fine_tuning.py:26]: aren't all tests executing the same logic?

File: test/renate/updaters/experimental/test_joint.py [hunk 7]
```diff
@@ -42,7 +42,7 @@ def test_joint_learner_model_reset():
     model_updater = pytest.helpers.get_simple_updater(
         model=model,
         learner_class=JointLearner,
-        learner_kwargs={"learning_rate": 0.0},
+        learner_kwargs={"learning_rate": 0.0, "loss_fn": torch.nn.CrossEntropyLoss()},
```
Comment (wistuba) [inline test/renate/updaters/experimental/test_joint.py:45]: again :)

File: test/renate/updaters/test_learner.py [hunk 8]
```diff
@@ -30,31 +28,34 @@ def get_model_and_learner_and_learner_kwargs(
...
+    learner.on_save_checkpoint(checkpoint=checkpoint_dict)
+    check_learner_variables(learner, checkpoint_dict)
 
 
 @pytest.mark.parametrize("learner_class", LEARNERS)
 def test_update_hyperparameters(learner_class):
     model, learner, learner_kwargs = get_model_and_learner_and_learner_kwargs(learner_class)
     check_learner_variables(learner, learner_kwargs)
-    learner.update_hyperparameters({})
+    learner.save_hyperparameters()
```
Comment (wistuba) [inline test/renate/updaters/test_learner.py:53]: what is this test doing?

File: test/renate/updaters/test_learner.py [hunk 9]
```diff
@@ -30,31 +28,34 @@ def get_model_and_learner_and_learner_kwargs(
...
     check_learner_variables(learner, learner_kwargs)
-    learner.update_hyperparameters(LEARNER_HYPERPARAMETER_UPDATES[learner_class])
     learner_kwargs = dict(learner_kwargs)
     learner_kwargs.update(LEARNER_HYPERPARAMETER_UPDATES[learner_class])
-    check_learner_variables(learner, learner_kwargs)
+    new_learner = learner_class(model=model, **learner_kwargs)
```
Comment (wistuba) [inline test/renate/updaters/test_learner.py:57]: and this one?

File: test/renate/updaters/test_model_updater.py [hunk 10]
```diff
@@ -204,6 +207,23 @@ def identity_transform():
...
+        learner_class=learner_class,
+        **transforms_kwargs,
+        learner_kwargs=LEARNER_KWARGS[learner_class],
     )
     check_learner_transforms(model_updater._learner, transforms_kwargs)
+
+
+@pytest.mark.xfail(raises=(KeyError, TypeError))
+@pytest.mark.parametrize("learner", LEARNERS_USING_SIMPLE_UPDATER)
+def test_learner_fails_without_loss_fn(learner):
```
Comment (wistuba) [inline test/renate/updaters/test_model_updater.py:221]: I don't think this test is particular useful. maybe you want to check if they use the loss instead?

File: src/renate/utils/deepspeed.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
+# Copyright 2020 The PyTorch Lightning team and Microsoft Corporation. All rights reserved.
```
Comment (wistuba) [inline src/renate/utils/deepspeed.py:1]: what is with this copyright note?

Comment (wistuba) [inline src/renate/utils/deepspeed.py:1]: this file in particular seems to need a bit more coverage

**Review by wistuba — COMMENTED** [review]
we could increase (meaningful) coverage a bit more. is the extra state conversion somehow covered?

missing documentation is provided in the next PR?

File: examples/nlp_finetuning/renate_config.py [hunk 12]
```diff
@@ -17,14 +18,20 @@ def model_fn(model_state_url: Optional[str] = None) -> RenateModule:
...
     return model
 
 
-def data_module_fn(data_path: str, chunk_id: int, seed: int = defaults.SEED) -> RenateDataModule:
+def loss_fn() -> torch.nn.Module:
+    return torch.nn.CrossEntropyLoss()
+
+
+def data_module_fn(
+    data_path: Union[Path, str], chunk_id: int, seed: int = defaults.SEED
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply examples/nlp_finetuning/renate_config.py:33]: Changing

File: doc/getting_started/how_to_renate_config.rst [hunk 13]
```diff
@@ -124,7 +124,7 @@ or created ad-hoc by implementing the same interface
 
 .. literalinclude:: ../../examples/getting_started/renate_config.py
     :caption: Example
-    :lines: 93-
+    :lines: 91-
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply doc/getting_started/how_to_renate_config.rst:127]: The main one yes.

File: requirements.txt [hunk 14]
```diff
@@ -12,3 +12,4 @@ torchmetrics~=0.10.3
 torchvision>=0.13.0, <0.15.2
 datasets~=2.9.0
 transformers>4.23.0, <4.26.2
+deepspeed==0.9.1
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply requirements.txt:15]: i haven't ran exps with that. hence this version.

File: src/renate/training/training.py [hunk 15]
```diff
@@ -550,7 +562,9 @@ def _execute_training_and_tuning_job_locally(
             f"Tuning hyperparameters with respect to {metric} ({mode}) for {max_time} seconds on "
             f"{n_workers} worker(s)."
         )
-    backend = LocalBackend(entry_point=training_script)
+    # TODO: After bumping up SyneTune >= 0.6, use the argument `num_gpus_per_trial`.
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/training/training.py:565]: its not supported before that.

File: src/renate/utils/distributed_strategies.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        raise ValueError("Number of devices has to be at least 0.")
+
+    elif devices == 1:
+        # If one GPU, use standard training. Enabled by passing strategy=None
+        # to pl.Trainer
+        if strategy_name is not None:
+            warnings.warn(f"With devices=1, strategy is ignored. But got {strategy_name}.")
+
+        return None
+    elif strategy_name in ["none", "None", None]:
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/utils/distributed_strategies.py:43]: Lightning has strategy as `Union[str, Strategy, None]`. So I guess this covers the common use cases.

File: src/renate/utils/distributed_strategies.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    elif "deepspeed" in strategy_name:
+        strategy = StrategyRegistry.get(strategy_name)
+
+        ## TODO: This should be changed to instantiating Deepspeed and settting it in
+        # the constructor. This works for nowbecause forcing PyTorch optimizer flag isn't used
+        # anywhere by Deepspeed.
+        strategy.config["zero_force_ds_cpu_optimizer"] = False
+        return strategy
+
+    else:
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/utils/distributed_strategies.py:56]: It crashes because of something in deepspeed. See https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L1164

File: src/renate/updaters/model_updater.py [hunk 18]
```diff
@@ -40,13 +43,14 @@ def _log(self, trainer: Trainer, training: bool) -> None:
...
 
-        if trainer.sanity_checking or (training and self._val_enabled):
-            return
-        self._report(
-            **{k: v.item() for k, v in trainer.logged_metrics.items()},
-            step=trainer.current_epoch,
-            epoch=trainer.current_epoch + 1,
-        )
+        if trainer.is_global_zero:
+            if trainer.sanity_checking or (training and self._val_enabled):
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/model_updater.py:47]: > missing documentation is provided in the next PR?

Yes.

[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/model_updater.py:47]: This is cleaner IMO. `Log if rank is zero` seems the obvious way to phrase it

File: src/renate/models/renate_module.py [hunk 19]
```diff
@@ -54,10 +55,10 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         loss_fn: The loss function to be optimized during the training.
     """
 
-    def __init__(self, constructor_arguments: dict, loss_fn: torch.nn.Module):
+    def __init__(self, constructor_arguments: dict):
         super(RenateModule, self).__init__()
         self._constructor_arguments = copy.deepcopy(constructor_arguments)
-        self.loss_fn = loss_fn
+        ## Loss function is a part of learner.
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/models/renate_module.py:61]: Probably, not.

File: src/renate/cli/parsing_functions.py [hunk 20]
```diff
@@ -286,6 +287,22 @@ def _standard_arguments() -> Dict[str, Dict[str, Any]]:
             "help": f"Devices used for this job. Default: {defaults.DEVICES} device.",
             "argument_group": OPTIONAL_ARGS_GROUP,
         },
+        "strategy": {
+            "type": str,
+            "default": defaults.DISTRIBUTED_STRATEGY,
+            "help": f"Distributed training strategy when devices > 1. Default:"
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/cli/parsing_functions.py:293]: I guess the second part needs it.

File: src/renate/utils/misc.py [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+    """Function to cast to int or str. This is used to tackle precision
+    which can be int (16, 32) or str (bf16)"""
+    try:
+        return int(x)
+    except ValueError:
+        return x
+
+
+@rank_zero_only
+def unlink_file_or_folder(path: Path) -> None:
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/utils/misc.py:19]: no longer a path? or `Union[Path, str]`?

File: src/renate/utils/deepspeed.py (hunk 11 above)
Comment (prabhuteja12) [reply src/renate/utils/deepspeed.py:1]: Its an adaptation of lightning utils file. So this has an original copyright, and modifications are according to our spec. See line 45. Should this be changed?

Comment (prabhuteja12) [reply src/renate/utils/deepspeed.py:1]: Unsure how.

File: test/conftest.py [hunk 22]
```diff
@@ -254,6 +270,11 @@ def get_renate_module_mlp(
     )
 
 
+@pytest.helpers.register
```
Comment (lballes) [inline test/conftest.py:273]: Seems overkill for a single line ;)

File: test/conftest.py [hunk 23]
```diff
@@ -338,6 +359,31 @@ def get_renate_module_mlp_and_data(
     return model, train_dataset, test_data
 
 
+@pytest.helpers.register
+def get_renate_module_mlp_data_and_loss(
```
Comment (lballes) [inline test/conftest.py:363]: I would replace the previous helper with this to always return the loss function alongside.

File: doc/getting_started/how_to_renate_config.rst
Comment (lballes) [inline doc/getting_started/how_to_renate_config.rst:1]: I think we also need to add to the doc page that users need to implement the `loss_fn`.

**Review by lballes — CHANGES_REQUESTED** [review]

File: src/renate/benchmark/models/base.py [hunk 24]
```diff
@@ -83,13 +82,20 @@ def get_params(self, task_id: str = defaults.TASK_ID) -> List[torch.nn.Parameter
             self.get_predictor(task_id=task_id).parameters()
         )
 
-    def get_extra_state(self) -> Any:
-        """Get the constructor_arguments, loss and task ids necessary to reconstruct the model."""
-        extra_state = super().get_extra_state()
+    def get_extra_state(self, encode=True) -> Any:
+        """Get the constructor_arguments, loss and task ids necessary to reconstruct the model.
+        Encode converts the state into a torch tensor so that Deepspeed serialization works.
+        We don't encoder any of the super() calls, but encode only the final dict."""
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/benchmark/models/base.py:88]: Added.

File: requirements.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply requirements.txt:15]: This has been tried out. I don't know of any breaking changes in other versions. What should it be?

File: src/renate/benchmark/models/base.py [hunk 25]
```diff
@@ -83,13 +82,20 @@ def get_params(self, task_id: str = defaults.TASK_ID) -> List[torch.nn.Parameter
             self.get_predictor(task_id=task_id).parameters()
         )
 
-    def get_extra_state(self) -> Any:
-        """Get the constructor_arguments, loss and task ids necessary to reconstruct the model."""
-        extra_state = super().get_extra_state()
+    def get_extra_state(self, encode=True) -> Any:
+        """Get the constructor_arguments, loss and task ids necessary to reconstruct the model.
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/benchmark/models/base.py:86]: Done

File: src/renate/defaults.py [hunk 26]
```diff
@@ -33,6 +33,8 @@
 SUPPORTED_ACCELERATORS_TYPE = Literal["auto", "cpu", "gpu", "tpu"]
 DEVICES = 1
 VOLUME_SIZE = 60
+DISTRIBUTED_STRATEGY = "ddp"
+PRECISION = "32"
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/defaults.py:37]: No. This is to simplify our `parsing_functions.py` to have type str. This is converted to `int` near the `Trainer` in `model_updater.py`

File: src/renate/training/training.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/training/training.py:565]: Yes #238

File: src/renate/updaters/learner.py [hunk 27]
```diff
@@ -92,6 +94,23 @@ def __init__(
         self._val_memory_buffer: DataBuffer = InfiniteBuffer()
         self._create_metrics_collections(logged_metrics)
         self._post_init()
+        self.save_hyperparameters(
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:97]: Is this for save_hyperparameters?

File: src/renate/updaters/learner.py [hunk 28]
```diff
@@ -92,6 +94,23 @@ def __init__(
...
+                "train_target_transform",
+                "test_transform",
+                "test_target_transform",
+                "buffer_transform",
+                "buffer_target_transform",
+                "logged_metrics",
+            ]
+        )
 
     def _post_init(self) -> None:
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:115]: This has moved. Can you clarify?

File: src/renate/updaters/learner.py [hunk 29]
```diff
@@ -411,32 +379,42 @@ def __init__(
         **kwargs,
     ) -> None:
         super().__init__(seed=seed, **kwargs)
+        self.save_hyperparameters(
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:378]: Only one (latest) hyperparameter is stored. See https://github.com/Lightning-AI/lightning/issues/16206 and https://github.com/Lightning-AI/lightning/blob/master/tests/tests_pytorch/models/test_hparams.py

File: src/renate/updaters/learner.py [hunk 30]
```diff
@@ -411,32 +379,42 @@ def __init__(
...
+                "test_transform",
+                "test_target_transform",
+                "buffer_transform",
+                "buffer_target_transform",
+                "logged_metrics",
+            ]
+        )
+
         self._memory_batch_size = min(memory_size, memory_batch_size)
+        self.hparams["memory_batch_size"] = self._memory_batch_size
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:397]: No. `memory_batch_size` is recomputed one line above. So that is updated in the `self.hparams` attribute.

File: src/renate/updaters/learner.py [hunk 31]
```diff
@@ -411,32 +379,42 @@ def __init__(
...
+        super().on_save_checkpoint(checkpoint)
+        checkpoint["memory_buffer"] = self._memory_buffer.state_dict()
 
-    def load_state_dict(self, model: RenateModule, state_dict: Dict[str, Any], **kwargs) -> None:
-        """Restores the state of the learner."""
-        super().load_state_dict(model, state_dict, **kwargs)
-        self._memory_batch_size = state_dict["memory_batch_size"]
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        super().on_load_checkpoint(checkpoint)
         if not hasattr(self, "_memory_buffer"):
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:411]: What can?

File: test/renate/updaters/experimental/test_fine_tuning.py (hunk 6 above)
Comment (prabhuteja12) [reply test/renate/updaters/experimental/test_fine_tuning.py:26]: This is only a sanity check to ensure that  all(-most) possible inputs work.

File: test/conftest.py (hunk 22 above)
Comment (prabhuteja12) [reply test/conftest.py:273]: It is reused subsequently.

File: test/renate/updaters/avalanche/test_avalanche_model_updater.py (hunk 4 above)
Comment (prabhuteja12) [reply test/renate/updaters/avalanche/test_avalanche_model_updater.py:29]: Done.

File: src/renate/cli/parsing_functions.py [hunk 32]
```diff
@@ -286,6 +287,22 @@ def _standard_arguments() -> Dict[str, Dict[str, Any]]:
...
+            + "{defaults.DISTRIBUTED_STRATEGY}.",
+            "argument_group": OPTIONAL_ARGS_GROUP,
+            "choices": list(pl.strategies.StrategyRegistry.keys()),
+        },
+        "precision": {
+            "type": str,
+            "default": defaults.PRECISION,
+            "help": f"Distributed training strategy when devices > 1."
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/cli/parsing_functions.py]: done.

File: src/renate/updaters/learner.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/updaters/learner.py:115]: This refers to the `_post_init` method. We initially had this method, because when recreating the `Learner` from a new state, we wouldn't call it's constructor. My understanding is that we now *do* call the constructor (via Lightning's reloading mechanism) so the stuff in `_post_init` could be moved to the regular `__init__`. Am I missing something?

File: src/renate/updaters/learner.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/updaters/learner.py:397]: But wouldn't the same recomputation happen upon recreating the object? (This is related to my question whether `LightningModule.from_checkpoint` calls the constructor).

File: src/renate/updaters/learner.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/updaters/learner.py:378]: Not sure I follow the discussion in the threads you referenced. But might it be safer to explicitly specify which hyperparameters to save, i.e., `save_hyperparameters("seed", ...)`? Might also be a bit shorter, given that we ignore a lot of them.

File: src/renate/updaters/learner.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/updaters/learner.py:411]: Can it still happen that the `Learner` does not have an attribute `"_memory_buffer"` when `on_load_checkpoint` is called? I think no.

File: src/renate/updaters/learner.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:378]: There are one too many to manually add (see SuperER). The ones we ignore are attributes read from the renate_config.py and components.

File: src/renate/updaters/learner.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:397]: Yes, that can be removed.

File: src/renate/updaters/learner.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:411]: It shouldn't because checkpoint loader calls the constructor. Removing this.

File: src/renate/updaters/learner.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:115]: If the reason for this function is the previous `__new__` call, then this can done away with. Should this cleanup be part of this PR or later?

File: src/renate/utils/misc.py [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+    """Funtion to remove files and folders. Unlink works for files, rmdir
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/utils/misc.py:20]: Changed

File: src/renate/utils/file.py [hunk 34]
```diff
@@ -277,3 +279,15 @@ def save_pandas_df_to_csv(df: pd.DataFrame, file_path: Union[str, Path]) -> pd.D
     """
     df.to_csv(file_path, index=False)
     return df
+
+
+def convert_to_tensor(obj):
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/utils/file.py:284]: Done. The main comment is no longer relevant here.

File: examples/nlp_finetuning/start.py [hunk 35]
```diff
@@ -44,4 +44,7 @@
         instance_count=1,
         instance_type="ml.g4dn.xlarge",
         job_name="renate-training-nlp-finetuning",
+        devices=1,
+        strategy="deepspeed_stage_2",
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply examples/nlp_finetuning/start.py:48]: Yes.

File: src/renate/models/renate_module.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/models/renate_module.py]: Done

File: src/renate/updaters/learner.py [hunk 36]
```diff
@@ -125,52 +144,18 @@ def _create_metrics_collections(
...
-        self._seed = state_dict["seed"]
-        self._task_id = state_dict["task_id"]
-        if not hasattr(self, "_val_memory_buffer"):
-            self._val_memory_buffer = InfiniteBuffer()
+    def on_load_checkpoint(self, state_dict):
         self._val_memory_buffer.load_state_dict(state_dict["val_memory_buffer"])
+        if "optimizer" in state_dict:
+            # It not deepspeed, we get this element. So we manually load that.
+            self._optimizer = state_dict["optimizer"]
         self._post_init()
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/updaters/learner.py]: No. Removed

File: examples/getting_started/renate_config.py [hunk 37]
```diff
@@ -15,9 +15,7 @@ class MyMNISTMLP(RenateModule):
     def __init__(self, num_hidden: int) -> None:
         # Model hyperparameters as well as the loss function need to registered via RenateModule's
```
Comment (lballes) [inline examples/getting_started/renate_config.py:16]: Remove reference to loss function in comment.

File: src/renate/updaters/experimental/er.py [hunk 38]
```diff
@@ -57,15 +57,30 @@ def __init__(
     ) -> None:
         self._components_names = list(components.keys())
         super().__init__(**kwargs)
+        self._memory_loader: Optional[DataLoader] = None  # previously post_init
```
Comment (lballes) [inline src/renate/updaters/experimental/er.py:60]: Remove comment.

File: src/renate/updaters/learner.py [hunk 39]
```diff
@@ -91,10 +93,24 @@ def __init__(
 
         self._val_memory_buffer: DataBuffer = InfiniteBuffer()
         self._create_metrics_collections(logged_metrics)
-        self._post_init()
-
-    def _post_init(self) -> None:
-        self._rng = get_generator(self._seed)
+        self._rng = get_generator(self._seed)  # previously post_init
```
Comment (lballes) [inline src/renate/updaters/learner.py:96]: Remove comment.

File: src/renate/updaters/learner.py [hunk 40]
```diff
@@ -125,53 +141,15 @@ def _create_metrics_collections(
             }
         )
 
-    def state_dict(self, **kwargs) -> Dict[str, Any]:
-        """Returns the state of the learner."""
-        return {
+    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        learner_state_dict = {
             "learner_class_name": self.__class__.__name__,
```
Comment (lballes) [inline src/renate/updaters/learner.py:146]: It seems like this is not used anymore. I don't mind leaving it in, but we could remove it. Your call.

File: src/renate/updaters/learner.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/updaters/learner.py:356]: I am not sure about this. Based on playing with a toy example, I think the `save_hyperparameters` call in the base class might be enough. Let's talk about it offline.

File: src/renate/updaters/model_updater.py [hunk 41]
```diff
@@ -125,15 +127,37 @@ def _load_best_checkpoint_and_save(self, trainer: Trainer, pl_module: LightningM
         # Reload best state.
         learner_state_path = Path(defaults.learner_state_file(self._output_state_folder))
         if learner_state_path.exists():
-            self._model.load_state_dict(torch.load(defaults.model_file(self.dirpath)))
-            pl_module.load_state_dict(self._model, torch.load(learner_state_path)["state_dict"])
+            loaded_state = trainer.strategy.load_checkpoint(learner_state_path)
+            pl_module.on_load_checkpoint(loaded_state)
+            # This loads the state dict only if its not Deepspeed.
```
Comment (lballes) [inline src/renate/updaters/model_updater.py:132]: Unsure what the comment refers to. Can you be more specific about which part is needed for Deepspeed?

File: src/renate/updaters/model_updater.py [hunk 42]
```diff
@@ -125,15 +127,37 @@ def _load_best_checkpoint_and_save(self, trainer: Trainer, pl_module: LightningM
...
+            trainer.strategy.load_model_state_dict(loaded_state)
+
         # Finalize model update.
         pl_module.on_model_update_end()
         # Save permanently.
         pl_module.save(self._output_state_folder)
         # Overwrite checkpoint.
         self._save_checkpoint(trainer, learner_state_path)
 
+    def teardown(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
```
Comment (lballes) [inline src/renate/updaters/model_updater.py:143]: A doc string would be helpful here. I don't exactly follow what we do here. Is the goal to have a separate file that just contains the model weights?

**Review by lballes — CHANGES_REQUESTED** [review]

**Comment by lballes** [comment]
We need to check the doc strings. I saw that the `RenateModule` doc string still contains the `loss_fn` argument. Can you make sure that all doc strings reflect the change? I.e., remove it from `RenateModule` and its subclasses and add it to `Learner` and its subclasses.

File: src/renate/updaters/learner.py (hunk 40 above)
Comment (prabhuteja12) [reply src/renate/updaters/learner.py:146]: it isnt. But I left it in as a sanity check.

File: src/renate/updaters/model_updater.py (hunk 42 above)
Comment (prabhuteja12) [reply src/renate/updaters/model_updater.py:143]: yes. added a detailed description

**Review by lballes — APPROVED** [review]

**Review by lballes — DISMISSED** [review]

**Review by lballes — APPROVED** [review]
